File: automations/src/trigger/brain/fetch-analytics.ts
import { task } from "@trigger.dev/sdk";

interface AnalyticsResult {
  source: string;
  period?: string;
  status?: string;
  message?: string;
  [key: string]: any;
}

async function fetchGA4(
  propertyId: string,
  token: string,
  days: number
): Promise<AnalyticsResult> {
  const end = new Date().toISOString().slice(0, 10);
  const start = new Date(Date.now() - days * 86400000).toISOString().slice(0, 10);

  const payload = {
    dateRanges: [{ startDate: start, endDate: end }],
    metrics: [
      { name: "activeUsers" },
      { name: "screenPageViews" },
      { name: "sessions" },
      { name: "bounceRate" },
    ],
  };

  const url = `https://analyticsdata.googleapis.com/v1beta/properties/${propertyId}:runReport`;
  const resp = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!resp.ok) {
    return {
      source: "ga4",
      status: "error",
      code: resp.status,
      message: await resp.text(),
    };
  }

  const data = await resp.json();
  const rows = data.rows || [];
  if (rows.length > 0) {
    const values = rows[0].metricValues || [];
    return {
      source: "ga4",
      period: `${start} to ${end}`,
      active_users: values[0]?.value ?? null,
      pageviews: values[1]?.value ?? null,
      sessions: values[2]?.value ?? null,
      bounce_rate: values[3]?.value ?? null,
    };
  }
  return { source: "ga4", period: `${start} to ${end}`, status: "no_rows" };
}

async function fetchLinkedIn(token: string): Promise<AnalyticsResult> {
  const url =
    "https://api.linkedin.com/v2/organizationalEntityShareStatistics?q=organizationalEntity";
  const resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });

  if (!resp.ok) {
    return {
      source: "linkedin",
      status: "error",
      code: resp.status,
      message: await resp.text(),
    };
  }

  return { source: "linkedin", data: await resp.json() };
}

export const fetchAnalytics = task({
  id: "fetch-analytics",
  run: async (payload: {
    source?: "ga4" | "linkedin" | "all";
    days?: number;
  }) => {
    const source = payload.source || "all";
    const days = payload.days || 7;
    const results: AnalyticsResult[] = [];

    if (source === "ga4" || source === "all") {
      const gaId = process.env.GOOGLE_ANALYTICS_ID;
      const gaToken = process.env.GOOGLE_ANALYTICS_TOKEN;
      if (gaId && gaToken) {
        results.push(await fetchGA4(gaId, gaToken, days));
      } else {
        results.push({
          source: "ga4",
          status: "skipped",
          message: "GOOGLE_ANALYTICS_ID or GOOGLE_ANALYTICS_TOKEN not set",
        });
      }
    }

    if (source === "linkedin" || source === "all") {
      const liToken = process.env.LINKEDIN_ACCESS_TOKEN;
      if (liToken) {
        results.push(await fetchLinkedIn(liToken));
      } else {
        results.push({
          source: "linkedin",
          status: "skipped",
          message: "LINKEDIN_ACCESS_TOKEN not set",
        });
      }
    }

    console.log(
      `Fetched analytics: ${results.map((r) => `${r.source}=${r.status || "ok"}`).join(", ")}`
    );
    return results;
  },
});

File: apex-brain/tools/fetch_analytics.py
#!/usr/bin/env python3
"""Fetch basic metrics from Google Analytics 4 and LinkedIn page analytics."""

import argparse
import json
import os
import sys
import urllib.request
import urllib.error
from datetime import datetime, timedelta


def fetch_ga4(property_id: str, token: str, days: int = 7) -> dict:
    """Pull pageviews and users from GA4 Data API."""
    end = datetime.utcnow().strftime("%Y-%m-%d")
    start = (datetime.utcnow() - timedelta(days=days)).strftime("%Y-%m-%d")
    payload = {
        "dateRanges": [{"startDate": start, "endDate": end}],
        "metrics": [
            {"name": "activeUsers"},
            {"name": "screenPageViews"},
            {"name": "sessions"},
            {"name": "bounceRate"},
        ],
    }
    url = f"https://analyticsdata.googleapis.com/v1beta/properties/{property_id}:runReport"
    body = json.dumps(payload).encode()
    req = urllib.request.Request(url, data=body, headers={
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
    })
    try:
        with urllib.request.urlopen(req) as resp:
            data = json.loads(resp.read())
        rows = data.get("rows", [])
        if rows:
            values = rows[0].get("metricValues", [])
            return {
                "source": "ga4",
                "period": f"{start} to {end}",
                "active_users": values[0]["value"] if len(values) > 0 else None,
                "pageviews": values[1]["value"] if len(values) > 1 else None,
                "sessions": values[2]["value"] if len(values) > 2 else None,
                "bounce_rate": values[3]["value"] if len(values) > 3 else None,
            }
        return {"source": "ga4", "period": f"{start} to {end}", "data": "no rows"}
    except urllib.error.HTTPError as e:
        return {"source": "ga4", "status": "error", "code": e.code, "message": e.read().decode()}


def fetch_linkedin(token: str) -> dict:
    """Pull LinkedIn organization page stats."""
    url = "https://api.linkedin.com/v2/organizationalEntityShareStatistics?q=organizationalEntity"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {token}"})
    try:
        with urllib.request.urlopen(req) as resp:
            return {"source": "linkedin", "data": json.loads(resp.read())}
    except urllib.error.HTTPError as e:
        return {"source": "linkedin", "status": "error", "code": e.code, "message": e.read().decode()}


def main():
    parser = argparse.ArgumentParser(description="Fetch analytics metrics")
    parser.add_argument("--source", choices=["ga4", "linkedin", "all"], default="all")
    parser.add_argument("--days", type=int, default=7, help="Lookback period for GA4")
    parser.add_argument("--dry-run", action="store_true", help="Show what would be fetched")
    args = parser.parse_args()

    if args.dry_run:
        print(json.dumps({"status": "dry_run", "source": args.source, "days": args.days}))
        return

    results = []
    if args.source in ("ga4", "all"):
        ga_id = os.environ.get("GOOGLE_ANALYTICS_ID")
        ga_token = os.environ.get("GOOGLE_ANALYTICS_TOKEN")
        if ga_id and ga_token:
            results.append(fetch_ga4(ga_id, ga_token, args.days))
        else:
            results.append({"source": "ga4", "status": "skipped", "message": "GOOGLE_ANALYTICS_ID or GOOGLE_ANALYTICS_TOKEN not set"})

    if args.source in ("linkedin", "all"):
        li_token = os.environ.get("LINKEDIN_ACCESS_TOKEN")
        if li_token:
            results.append(fetch_linkedin(li_token))
        else:
            results.append({"source": "linkedin", "status": "skipped", "message": "LINKEDIN_ACCESS_TOKEN not set"})

    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    main()
